### ISSUE TITLE
fish unit tests involving unicode fail on Cygwin (MS Windows)

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#3044]
Running `make test` on Cygwin resulted in test failures such as

```
Testing utils
Error: u2w: resulted in forbitten wchars (len): FAILED (rv: 3, must be 0)
```

That made we wonder why `utf8_to_wchar()` exists. As far as I can tell it is used in exactly two places (directly or indirectly):

1) *src/env_universal_common.cpp*

2) *src/fish_tests.cpp*

It seems to me it should be removed.

**Comment by krader1961** [comment]
We're going to need something like this function as part of my proposal to use UTF-8 narrow strings internally. So I'm going to replace it with functions named something like `utf8_to_locale` and `locale_to_utf8` and add tests for the new functions.

**Comment by ridiculousfish** [comment]
This function is used when reading from the universal variables file. What would the difference be between `utf8_to_wchar` and `utf8_to_locale`?

**Comment by krader1961** [comment]
Sorry, my previous comment doesn't make any sense. Obviously the existing `utf8_to_wchar()` and `wchar_to_utf8()` functions must continue to exist until fish is using UTF-8 encoded narrow strings internally. We need to be able to read and write the universal file contents without regard to the user's locale. Once fish is using UTF-8 internallly those functions are no longer needed and can be deleted.

I've been working on making `make test` pass on Cygwin and I'm appalled at the number of test failures. All of the failures appear to involve UTF-8 to wide char, or vice-versa, conversions. And those failures are, not surprisingly, because parts of the core fish code and the unit tests assume a **wchar_t** is four, not two, bytes in size. There are some attempts to accommodate a two byte size (e.g., via the `is_wchar_ucs2()` function) but they're mostly broken. The UTF-8 conversion code and associated tests also try to use five and six byte UTF-8 sequences that were specified by the original Unicode proposal but are not part of the standard (see https://en.wikipedia.org/wiki/UTF-8).

> What would the difference be between utf8_to_wchar and utf8_to_locale?

The former is used when reading universal vars. The latter would be used when writing (or putting into the environment) internal UTF-8 strings so they are in the char set expected by the user's environment. Thus they're not analogs of each other. If the user's environment is a UTF-8 locale it is a no-op.

**Comment by krader1961** [comment]
Related to my previous comment: The `test_convert()` function fails on Cygwin with this UTF-8 sequence: `0x55 0x67 0x97 0xF3 0xA4 0x8E` (a random string from *fish_tests.cpp*). That is because the 0xF3 char introduces a four byte UTF-8 sequence but it's invalid because it is only three bytes in length. Thus the `str2wcstring()` and `wcs2str()` are not lossless with respect to four, five and six byte UTF-8 sequences on Cygwin.

Since we don't seem to have had a lot of Cygwin users complain about our Unicode support I can only conclude that very few fish users are using Asian char sets not in the BMP.

**Comment by krader1961** [comment]
Sigh. The `str2wcs_internal()` function also assumes a **wchar_t** is four bytes in length. Have I mentioned recently how much I despise MS Windows? It's not just that it isn't UNIX or has a few bogosities. It's that it has numerous bogosities which make a software engineer scratch their head in puzzlement as to what the creators of MS Windows were thinking.

This wouldn't be an issue, in most cases, if fish used UTF-8 narrow strings internally rather than sequences of "wide chars".

**Comment by krader1961** [comment]
The Cygwin `mbrtowc()` function is converting 0xF3 0xA4 0x8E  to 0xDB50, a surrogate code point, rather than returning -2 as it should to indicate an incomplete sequence. Googling didn't turn up a definitive answer but I did find several mailing list messages (from 2009) that suggested this is a bug but may not be fixable because it relies on MS Windows functions for part of the implementation. I'm using a Windows 7 system so it's possible this is corrected on newer versions. Guess it's time to buy a copy of Windows 10.

**Comment by krader1961** [comment]
I bought a copy of MS Windows 10 and installed Cygwin on it. The Unicode related tests fail in the same manner seen on my MS Windows 7 system that has Cygwin installed. So either Cygwin and/or MS Windows is broken.

I'm going to modify `make test` to skip those tests on Cygwin. Frankly, it's not worth my time to ensure fish under Cygwin correctly handles all Unicode characters. I encourage someone who uses fish on Cygwin to investigate this problem and properly fix it.

**Comment by krader1961** [comment]
Running `./fish_tests` (or `make test`) on the Windows 10 Linux subsystem does not exhibit any wide-char, unicode, failures. It does fail the universal variable notifier tests just like happens on the DragonFly (a FreeBSD variant) OS that was mentioned in issue #3050. So, this is pretty clearly a bogosity of Cygwin on MS Windows.

I was thinking about adding code to *common.cpp* to special case Cygwin so that the tests would pass. Instead I've decided to simply skip the failing tests since no one has complained about character encoding issues on Cygwin and the problems involve handling of invalid UTF-8 sequences.

**Comment by ridiculousfish** [comment]
A wchar_t is supposed to be able to represent any character (by which I mean code point). I don't know if current Cygwin uses either UCS-2 or UTF-16, but either way it can't represent astral characters in a single wchar_t. So it's not surprising these tests fail. Just another reason to get off of wchar_t.

**Comment by krader1961** [comment]
@ridiculousfish: You're right. There is basically no sane way to handle non-BMP characters in a system where wchar_t is two bytes as on Cygwin. Given the needs of fish the cleanest way to solve this is to use fish's special encoding for UTF-8 sequences that require four bytes. I intend to do that rather than just disable the unit tests in that environment. That way we still verify that most characters are correctly handled on Cygwin and non-BMP UTF-8 sequences aren't silently discarded -- they'll be passed through fish without being recognized as characters by fish.

**Comment by krader1961** [comment]
I've corrected several problems with the *utf8.cpp* and *fish_tests.cpp* modules vis-a-vis Cygwin. There were problems unique to Cygwin such as variable initializations that don't produce the desired result due to narrowing conversions (e.g., assigning -2 to a wchar_t). There were also problems that theoretically affect UNIX systems where sizeof wchar_t is four. For example, the code in *utf8.cpp* allowing five and six byte UTF-8 sequences and associated tests in *fish_tests.cpp*.

The unicode tests (e.g., `./fish_tests utf8`) now pass. However, many other tests still fail. For example, `./fish-tests expand` fails with a segmentation fault. I'm mildly surprised that people using fish on Cygwin aren't reporting more problems.

**Comment by krader1961** [comment]
Fixed by commit 8f420b9272e526097a73ad10ce74ce3edbb67b9e. I'm so tired by fixing the Cygwin unicode tests that I'm going to defer the other test failures. Given the new Windows 10 "linux subsystem" it may be that we never spend the time to fix the other issues with fish on Cygwin.
